code/modules/deathmatch/deathmatch_lobby.dm

```diff
@@ -0,0 +1,436 @@
+
+	// Remove rest of spawns.
+	for (var/unused_spawn in spawns)
+		qdel(unused_spawn)
+
+	for (var/observer_key in observers)
+		var/mob/observer = observers[observer_key]["mob"]
+		observer.forceMove(location.loc)
+
+	addtimer(CALLBACK(src, PROC_REF(game_took_too_long)), initial(map.automatic_gameend_time))
+	log_game("Deathmatch game [host] started.")
+	announce(span_reallybig("GO!"))
+	return TRUE
+
+/datum/deathmatch_lobby/proc/spawn_observer_as_player(ckey, loc)
+	var/mob/dead/observer/observer = players[ckey]["mob"]
+	if (isnull(observer) || !observer.client)
+		remove_ckey_from_play(ckey)
+		return
+
+	// equip player
+	var/datum/outfit/deathmatch_loadout/loadout = players[ckey]["loadout"]
+	if (!(loadout in loadouts))
+		loadout = loadouts[1]
+
+	observer.forceMove(loc)
+	var/datum/mind/observer_mind = observer.mind
+	var/mob/living/observer_current = observer.mind?.current
+	var/mob/living/carbon/human/new_player = observer.change_mob_type(/mob/living/carbon/human, delete_old_mob = TRUE)
+	if(!isnull(observer_mind) && observer_current)
+		new_player.AddComponent( \
+			/datum/component/temporary_body, \
+			old_mind = observer_mind, \
+			old_body = observer_current, \
+		)
+	new_player.equipOutfit(loadout) // Loadout
+	players[ckey]["mob"] = new_player
+
+	// register death handling.
+	RegisterSignals(new_player, list(COMSIG_LIVING_DEATH, COMSIG_MOB_GHOSTIZED, COMSIG_QDELETING), PROC_REF(player_died))
+	if (global_chat)
+		ADD_TRAIT(new_player, TRAIT_SIXTHSENSE, INNATE_TRAIT)
+		ADD_TRAIT(new_player, TRAIT_XRAY_HEARING, INNATE_TRAIT)
+
+/datum/deathmatch_lobby/proc/game_took_too_long()
+	if (!location || QDELING(src))
+		return
+	announce(span_reallybig("The players have took too long! Game ending!"))
+	end_game()
+
+/datum/deathmatch_lobby/proc/end_game()
+	if (!location)
+		CRASH("Location of deathmatch game [host] deleted during game.")
+	var/mob/winner
+	if(players.len)
+		var/list/winner_info = players[pick(players)]
+		if(!isnull(winner_info["mob"]))
+			winner = winner_info["mob"] //only one should remain anyway but incase of a draw 
+	
+	announce(span_reallybig("THE GAME HAS ENDED.<BR>THE WINNER IS: [winner ? winner.real_name : "no one"]."))
+	
+	for(var/ckey in players)
+		var/mob/loser = players[ckey]["mob"]
+		UnregisterSignal(loser, list(COMSIG_MOB_GHOSTIZED, COMSIG_QDELETING))
+		players[ckey]["mob"] = null
+		loser.ghostize()
+		qdel(loser)
+
+	game.clear_location(location)
+	game.remove_lobby(host)
+	log_game("Deathmatch game [host] ended.")
+
+/datum/deathmatch_lobby/proc/player_died(mob/living/player, gibbed)
+	SIGNAL_HANDLER
+	if(isnull(player) || QDELING(src))
+		return
+
+	var/ckey = player.ckey
+	if(!islist(players[ckey])) // potentially the player info could hold a reference to this mob so we can figure the ckey out without worrying about ghosting and suicides n such
+		for(var/key in players)
+			var/list/player_info = players[key]
+			if(player_info["mob"] && player_info["mob"] == player)
+				ckey = key
```

Comment:
`ckey = key` generally seems incorrect, because ckey and key are two different forms, and shouldn't be intermixed